theano/sandbox/cuda/opt.py

```diff
@@ -2399,6 +2399,34 @@ def local_gpu_extract_diagonal(node):
     return False
 
 
+@register_opt()
+@local_optimizer([theano.tensor.Diagonal])
```

Comment:
You can add Diagonal to the optimization above and replace nlinalg.extract_diag by tensor.diagonal once you have modified Diagonal.perform method.

You should also add a test to make sure ExtractDiag and Diagonal behaves the same way.